drivers/clock_control/clock_control_npcx.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2020 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_pcc
+
+#include <soc.h>
+#include <drivers/clock_control.h>
+#include <dt-bindings/clock/npcx_clock.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(clock_control_npcx, LOG_LEVEL_ERR);
+
+/* Driver config */
+struct npcx_pcc_config {
+	/* cdcg controller base address */
+	cdcg_inst_t base_cdcg;
+	/* pmc controller base address */
+	pmc_inst_t base_pmc;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct npcx_pcc_config *)(dev)->config_info)
+
+#define HAL_CDCG_INST(dev) \
+	(cdcg_inst_t)(DRV_CONFIG(dev)->base_cdcg)
+
+#define HAL_PMC_INST(dev) \
+	(pmc_inst_t)(DRV_CONFIG(dev)->base_pmc)
+
+/******************************************************************************/
+/* Clock controller local functions */
+static inline int npcx_clock_control_on(struct device *dev,
+					 clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	struct npcx_clk_cfg *clk_cfg = (struct npcx_clk_cfg *)(sub_system);
+	pmc_inst_t inst_pmc = HAL_PMC_INST(dev);
+
+	/* Clear related PD (Power-Down) bit of module to turn on clock */
+	CLEAR_BIT(NPCX_PWDWN_CTL(inst_pmc, clk_cfg->ctrl), clk_cfg->bit);
```

Comment:
From my other comment in the nuvton_hal repository, I think this is were we want to have a device read/write funtion that uses the devices base address that is passed in as a parameter here instead of using hardcoded absolute locations. (The hard-coded offsets within a device still make sense though)

Reply:
Using relative offsets from a device base address (as defined in the device tree) seems like the way we should go instead of using absolute offsets? Is this something we can change to?